### DISCUSSION TITLE
Pull over push

### DISCUSSION THREAD

**Discussion by 6be709c0** [ethereum/ethereum-org-website#10599] [Q&A]
Hi,

I would like to know if pull over push is still a thing in Ethereum security on Layer 1 and 2.

As per an article from [openzeppelin](https://blog.openzeppelin.com/onward-with-ethereum-smart-contract-security-97a827e47702) in 2016 >
> Every Ether transfer implies potential execution. The receiving address can implement a [Fallback function](https://solidity.readthedocs.io/en/latest/contracts.html#fallback-function) that can throw an error. Thus, we should never trust that a send call will execute without error. A solution: our contracts should [favor pull over push for payments](https://github.com/ethereum/wiki/wiki/Safety#favor-pull-over-push-for-external-calls).

Which redirect to an old ethereum wiki, I can't find any information on the new wiki about pull over push and I am wondering if it's because there's a new way today or is it named differently ?

**Comment by Scottie33Pippen** [comment]
I'm interested, I'll take a look

**Comment by sohailblockchain** [comment]
Great question! Yes, the **Pull over Push pattern is absolutely still relevant** 
in Ethereum security for both Layer 1 and Layer 2 in 2024.

## Why It's Still Important:

The pull over push pattern remains a critical security best practice because it:

1. **Prevents Reentrancy Attacks** - By having users withdraw funds themselves 
   rather than the contract sending them, you eliminate reentrancy vulnerabilities

2. **Avoids Denial of Service** - If one recipient's fallback function fails, 
   it won't block other users from receiving their funds

3. **Gas Efficiency** - Users pay for their own withdrawal gas costs

## Current Documentation:

The pattern is still documented in modern resources:

- **OpenZeppelin Contracts** (current version) still uses this pattern in their 
  `PullPayment.sol` contract: 
  https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/PullPayment.sol

- **Consensys Smart Contract Best Practices** recommends this pattern:
  https://consensys.github.io/smart-contract-best-practices/attacks/denial-of-service/#favor-pull-over-push-for-external-calls

- **Solidity Documentation** mentions it in security considerations

## Layer 2 Relevance:

This pattern is equally important on Layer 2s (Optimism, Arbitrum, etc.) because:
- Same security vulnerabilities exist
- Gas considerations may differ but security principles remain
- Smart contract execution model is the same

## Example Pattern:
```solidity
